File: scripts/gripper_switch_server.py
#!/usr/bin/env python
import rospy
import sys
from std_msgs.msg import String
from ur_msgs.msg import IOStates
from magnetic_valve_switch.srv import switcher, switcherRequest,switcherResponse
import ipdb

_rate = 10

Received_Flag = None
Gripper_state = None

def callback(msg):
    global Received_Flag 
    Received_Flag = msg.digital_out_states[0].state

def service_cb(req):
    global Gripper_state
    res = switcherResponse()
    Gripper_state = req.gripper_state
    res.success = True
    return res

def main():
    rospy.init_node("io_swicth", anonymous=True)
    pub = rospy.Publisher('ur_driver/URScript', String, queue_size=1)
    rospy.wait_for_message('/ur_driver/io_states',IOStates,timeout=5)
    sub = rospy.Subscriber("/ur_driver/io_states", IOStates, callback)
    service = rospy.Service('gripper_switcher', switcher, service_cb)
    
    rate = rospy.Rate(_rate)
    while not rospy.is_shutdown():
        if Gripper_state == "open":
            while not rospy.is_shutdown():
                pub_msg = String()
                pub_msg.data = "set_digital_out(0,True)"
                pub.publish(pub_msg)
                rate.sleep()
                if Received_Flag == True:
                    break

        elif Gripper_state == "close":
            while not rospy.is_shutdown():
                pub_msg = String()
                pub_msg.data = "set_digital_out(0,False)"
                pub.publish(pub_msg)
                rate.sleep()
                if Received_Flag == False:
                    break


if __name__ == '__main__':
    try:
        sys.exit(main())   
    except rospy.ROSInterruptException:
        pass
        

File: src/gripper_switch_server.cpp
#include<ros/ros.h>
#include<iostream>
#include<std_msgs/String.h>
#include<ur_msgs/IOStates.h>
#include<magnetic_valve_switch/switcher.h>
using namespace std;

bool IO_flag;
string Gripper_state_flag;

void io_callback(const ur_msgs::IOStates::ConstPtr& msg)
{   
    IO_flag = msg->digital_out_states[0].state;
}

bool switcher_cb(magnetic_valve_switch::switcher::Request  &req, magnetic_valve_switch::switcher::Response &res)
{
    Gripper_state_flag = req.gripper_state;
    cout<< "Gripper_state_flag : " << Gripper_state_flag << endl;
    return true;
}

int main(int argc, char** argv)
{
    ros::init(argc, argv, "gripper_switch");
    ros::NodeHandle nh;
    ros::Rate loop_rate(10);
    ros::Publisher ur_script_pub;
    ros::Subscriber ur_state_sub;
    ros::ServiceServer gripper_state_service;

    ur_script_pub = nh.advertise<std_msgs::String>("ur_driver/URScript",1);
    ros::topic::waitForMessage<ur_msgs::IOStates>("/ur_driver/io_states", ros::Duration(10));
    ur_state_sub = nh.subscribe("/ur_driver/io_states", 100, io_callback);
    gripper_state_service = nh.advertiseService("gripper_switcher", switcher_cb);
    
    while(ros::ok())
    {
        if(Gripper_state_flag=="open")
        {   
            while(ros::ok())
            {  
                std_msgs::String cmd;
                cmd.data="set_digital_out(0,True)";
                ur_script_pub.publish(cmd);
                if(IO_flag == 1)
                {   
                    break;
                }
                ros::spinOnce();
                loop_rate.sleep();
            }
        }

        if(Gripper_state_flag=="close")
        {
            while(ros::ok())
            {   
                std_msgs::String cmd;
                cmd.data="set_digital_out(0,False)";
                ur_script_pub.publish(cmd);
                if(IO_flag == 0)
                {   
                    break;
                }
                ros::spinOnce();
                loop_rate.sleep();
            }
        }
    ros::spinOnce();
    loop_rate.sleep();
    }
    return 0;
}
